File: Submissions/343_Integer_Break.cpp
class Solution {
public:
    int integerBreak(int n) {
        vector<int> dp = {0, 1, 1, 2};
        
        for(int i=4; i<=n; ++i){
            int current = 0;
            for(int j=1; j<=i/2; ++j){
                current = max({current, dp[i-j]*dp[j], dp[i-j]*j, (i-j)*dp[j], (i-j)*j});
            }
            dp.push_back(current);
        }
        
        return dp[n];
    }
};

File: Kotlin/343_Integer_Break.kt
class Solution {
    fun integerBreak(n: Int): Int {
        if (n == 2 || n == 3) return n - 1
        
        val dp = IntArray(n + 1)
        dp[0] = 1
        dp[1] = 1
        for(i in 2..n){
            for(j in 2..i){
                dp[i] = maxOf(dp[i], dp[i-j] * j)
            }
        }
        
        return dp[n]
    }
}